Fixed #26891 -- Fixed lookup registration for ForeignObject

rpkilby opened the pull request:
https://code.djangoproject.com/ticket/26891

Update - just recapping the current state of the PR:

### Implementation:
The scope has expanded a bit. In fixing the original lookup registration issue, it was necessary to add `get_lookups()` to the registration API.

- `get_lookups()` simply returns the merged set of `class_lookups` in a field's class hierarchy. 
- For performance reasons, the lookups are cached on the field as `cached_lookups`. 
- Similar to registering lookups, `cached_lookups` are checked on the `cls.__dict__`. This prevents DateTimeField from accidentally using `Field.cached_lookups`, which wouldn't include the date transforms.
- `register_lookup()` will bust the cached lookups on a field and any of its subclasses. e.g., `ForeignObject.register_lookup()` would bust the cache of ForeignObject, ForeignKey, OneToOneField, etc...
- `ForeignObject.get_lookups()` only caches `class_lookups` up to it's position in the class hierarchy, excluding the lookups registered to Field.

### Performance tests:
Used timeit to time `field.get_lookup('in')` for different fields, comparing the current master to the PR. 100,000 iterations per run, 5 runs per field.

**master branch:**
```
CharField       average: 0.2338s [0.2377979755401611, 0.2334151268005371, 0.2305140495300293, 0.2334628105163574, 0.2340850830078125]
DateTimeField   average: 0.4872s [0.4839980602264404, 0.4864580631256103, 0.4956321716308594, 0.4867660999298095, 0.4834890365600586]
ForeignKey      average: 0.0266s [0.0261549949645996, 0.0266189575195312, 0.0267159938812255, 0.0269579887390136, 0.0267570018768310]
```

- Getting lookups on ForeignKey is fast because the lookups are hardcoded (which causes lookup registration to have no effect).
- DateTimeField is slower than CharField since it needs to test multiple `class_lookups` dicts.

**pr:**
```
CharField       average: 0.2818s [0.2849998474121094, 0.2799849510192871, 0.2836630344390869, 0.2778408527374267, 0.2827849388122558]
DateTimeField   average: 0.2833s [0.2864518165588379, 0.2799329757690429, 0.2777631282806396, 0.2892100811004638, 0.2832741737365722]
ForeignKey      average: 0.2851s [0.2807130813598633, 0.2836570739746094, 0.2891471385955810, 0.2907619476318359, 0.2814109325408935]
```

timgraham commented:
buildbot, add to whitelist.

timgraham commented:
Thanks for the patch. All non-trivial pull requests also require a Trac ticket.

rpkilby commented:
will do - should I just edit the PR once the Trac ticket is open?

timgraham commented:
Yep

rpkilby commented:
done

timgraham commented on django/db/models/fields/related.py:
```diff
@@ -731,26 +732,23 @@ def get_reverse_path_info(self):
+                # Related lookups need to be explicitly registered to related fields.
+                if parent is ForeignObject:
+                    break
+                if 'class_lookups' not in parent.__dict__:
```
Did you consider trying to make `RegisterLookupMixin._get_lookup()` more extensible so it's not required to copy/paste the whole thing to add the `if parent...` condition. This adds more untested code paths. For example, comment this if statement out doesn't cause any test failures.

rpkilby replied:
Two ideas:
- Add `bases` argument to `_get_lookup()`. This might be a good short term solution, since it only affects an internal method. 
  ```py
  # RegisterLookupMixin
      def _get_lookup(self, lookup_name, bases=None):
          ...
          if bases is None:
              bases = inspect.getmro(self.__class__)
          for parent in bases:
              ...

  # ForeignObject
      def _get_lookup(self, lookup_name, bases=None):
          if bases is None:
              bases = inspect.getmro(self.__class__)
              bases = bases[:bases.index(ForeignObject)+1]
          return super(ForeignObject, self)._get_lookup(lookup_name, bases)
  ```
- Add `RegisterLookupMixin.get_lookups()`. I kind of like this approach because it makes introspecting available lookups & transforms easier. It might also make `_get_lookup()` unnecessary.
  ```py
  def merge_dicts(dicts):
      merged = {}

      # merge in reverse to preference order
      for d in reversed(dicts):
          merged.update(d)

      return merged

  # RegisterLookupMixin
      def get_lookups(self):
          class_lookups = [getattr(cls, 'class_lookups', {}) for cls in inspect.getmro(self.__class__)]
          return merge_dicts(class_lookups)

      def _get_lookup(self, lookup_name):
          return self.get_lookups().get(lookup_name, None)

# ForeignObject
      def get_lookups(self):
          bases = inspect.getmro(self.__class__)
          bases = bases[:bases.index(ForeignObject)+1]
          class_lookups = [getattr(cls, 'class_lookups', {}) for cls in bases]
  ```

jarshwah replied:
I like your option 2 of `get_lookups`, I think it improves clarity over what we already have. You could also introduce some caching within the `get_lookups` to improve perf for classes that are looked up multiple times.

rpkilby replied:
Any recommendations or examples to work from? I ran a simple timeit test, and it looks like `get_lookups` performs worse (which makes sense). 


```
s = "models.DataPoint._meta.get_field('name').get_lookup('in')"
t = "models.RelatedPoint._meta.get_field('data').get_lookup('in')"

get_lookups:
s result [0.7386798858642578, 0.736055850982666, 0.7420699596405029, 0.7442851066589355, 0.7473750114440918]
t result [0.6083571910858154, 0.5898189544677734, 0.6105849742889404, 0.6074070930480957, 0.5925500392913818]

bases:
s result [0.2692680358886719, 0.2704458236694336, 0.2689228057861328, 0.2720210552215576, 0.2666950225830078]
t result [0.40657591819763184, 0.3987290859222412, 0.4173128604888916, 0.39420008659362793, 0.3984808921813965]
```

jarshwah replied:
Thanks for running some timing on these. We probably don't focus on performance as much as we should.

That's a fairly significant increase in time for the non related field (nearly 3 times slower). 50% slower for related field is not great either. I'm not sure that caching the bases for each field type is something we can do without busting the cache in register_lookup. I haven't put enough thought into how this might work to offer a great solution.

What I was thinking of was something like:

```
def get_lookups(self):
        cached = getattr(self, 'cached_lookups')
        if cached is not None: return cached
        class_lookups = [getattr(cls, 'class_lookups', {}) for cls in inspect.getmro(self.__class__)]
        merged = merge_dicts(class_lookups)
        self.cached_lookups = merged
        return merged
```

Would have to be careful that this code cached on the correct class though, which is probably unlikely considering the placement of RegisterLookupMixin to the left.

What I was hoping to avoid was recursively searching the base hierarchy for every access of a lookup.

```Model.objects.filter(intfield__gt=5, intfield__lt=10)``` would search twice. It'd be a lot nicer if the second lookup (`lt`) was an index into a saved dict rather than another base search.

If there's no obvious way to add caching then I think it best to go with the quicker solution (assuming the quicker was your option 1 and not just the status quo.

@carljm I know you're working with performance data lately - have you got any comments here? Do you have any idea if this kind of overhead would be noticeable in big deployments?

rpkilby replied:
So I reran the perf tests on the current master and on a get_lookups() branch w/ caching. 

```
s = "models.DataPoint._meta.get_field('name').get_lookup('in')"
r = "models.RelatedPoint._meta.get_field('data').get_lookup('in')"

Django master:
s result [0.2662360668182373, 0.27753305435180664, 0.2679319381713867, 0.2692220211029053, 0.2688179016113281]
r result [0.058827877044677734, 0.057289838790893555, 0.05742311477661133, 0.05547499656677246, 0.05588197708129883]

get_lookups() w/ cache:
s result [0.3323400020599365, 0.34210205078125, 0.33652591705322266, 0.3279271125793457, 0.3341038227081299]
r result [0.3339991569519043, 0.3370540142059326, 0.33731698989868164, 0.3404569625854492, 0.3389909267425537]

get_lookups():
s result [0.7437207698822021, 0.7416818141937256, 0.7478539943695068, 0.7437272071838379, 0.7346601486206055]
r result [0.589561939239502, 0.5912292003631592, 0.5971081256866455, 0.5899128913879395, 0.5951800346374512]
```

A couple of notes:
- The cached lookups aren't significantly slower, and performance seems consistent across standard fields and related fields. 
- It's worth noting that the related lookups are fast on the current master because they're hardcoded.
- Caching works by adding a `cached_lookups` to the current class `__dict__`. 
- Cache busting works by simply `del`eting the `cached_lookups`.  It's necessary to do this to the subclasses too.
- I don't know if it's acceptable to use `__subclasses__()` in the cache busting method, however I don't know of another way of getting all subclasses for a Field.

jarshwah replied:
Looks good! The one concern I have here (I haven't tested it) is that caching will happen on the RegisterLookupMixin (or any other mixin that might be shared) and will return the wrong results when get_lookups is called on another class. Subclasses may also present a problem - say with DateTimeField and DateField. Perhaps the reverse merge_dict takes care of this?

rpkilby replied:
> The one concern I have here (I haven't tested it) is that caching will happen on the RegisterLookupMixin

Caching and checking on the `cls.__dict__` should ensure that we don't interact with cached lookups on parent classes (eg, on `RegisterLookupMixin`). Added a test to demonstrate.

timgraham commented:
Any ideas about this, @akaariai or @jarshwah?

jarshwah commented:
Looks mostly good to me. I'm not very familiar with `ForeignObject` though so I might have missed something obvious. Was there a particular lookup you were trying to create that led you to fixing this? The test case you've got (`Exactly`) shows the feature working, but not really why the changed needed to be made.

rpkilby commented:
Hi all - I've recapped the PR in the description. Is there anything I can do to help move the PR forward?

jarshwah commented on django/db/models/query_utils.py:
```diff
@@ -27,6 +27,26 @@ class InvalidQuery(Exception):
+    # merge_dicts([a, b]) will preference the keys in a over those in b.
+    for d in reversed(dicts):
+        merged.update(d)
+
```
Remove this and the previous blank line to compress it all. Same goes for the subclasses method below. This is just a matter of taste and is entirely subjective.

rpkilby commented:
The blank lines have been removed.

My thoughts on updating the docs would be if `get_lookups()` was to be considered as public API or not.

timgraham commented on django/db/models/fields/related.py:
```diff
@@ -731,26 +732,15 @@ def get_reverse_path_info(self):
-            raise TypeError('Related Field got invalid lookup: %s' % lookup_name)
+    @classmethod
+    def get_lookups(cls):
+        if 'cached_lookups' not in cls.__dict__:
```
Could it be useful to use `@functools.lru_cache()` or `@cached_property` rather than implementing your own caching?

rpkilby replied:
The caching here is based on the same idea as `class_lookups`, just a bit more complicated because of the cache busting. I didn't try `@cached_property` (no way to bust the cached value, and it doesn't work w/ classmethods, right?), but `lru_cache` worked just fine. That said, it is about 50% slower.
  - current implementation at 100k runs takes ~.28s on average.
  - `lru_cache` at 100k runs takes ~.42s on average.

50% is a significant percentage, but we're talking about lookup resolution taking .0000042 seconds vs .0000028 seconds. `¯\_(ツ)_/¯`

Either way, I've squashed the existing commits into one, and added the lru changes separately. Thoughts on the performance vs code clarity?

timgraham commented on django/db/models/query_utils.py:
```diff
@@ -27,6 +27,23 @@ class InvalidQuery(Exception):
     pass
 
 
+def merge_dicts(dicts):
```
This doesn't strike me as an appropriate place given it's not reused anywhere. I'd put it in the file in which it's used, even possibly as a staticmethod next to the method in the class where it's used.

timgraham commented on django/db/models/query_utils.py:
```diff
@@ -27,6 +27,23 @@ class InvalidQuery(Exception):
 
+def merge_dicts(dicts):
+    merged = {}
+    # merge in reverse to preference the order of the original list. e.g.,
```
Make this a docstring?

timgraham commented on django/db/models/query_utils.py:
```diff
@@ -27,6 +27,23 @@ class InvalidQuery(Exception):
+
+def subclasses(cls):
+    yield cls
+    # python 2 lacks 'yield from'
```
Python (capitalize)
Can you expand the comment with what this should be changed to when dropping Python 2 compatibility?

timgraham commented on django/db/models/query_utils.py:
```diff
@@ -166,12 +181,19 @@ def get_transform(self, lookup_name):
         return found
 
     @classmethod
+    def _bust_cached_lookups(cls):
```
bust -> clear

timgraham commented:
I can't think of any docs that are needed.

timgraham commented:
You can squash commits when updating also. Thanks!

rpkilby commented:
Copying here since the diff is outdated:

----

> Could it be useful to use `@functools.lru_cache()` or `@cached_property` rather than implementing your own caching?

The caching here is based on the same idea as `class_lookups`, just a bit more complicated because of the cache busting. I didn't try `@cached_property` (no way to bust the cached value, and it doesn't work w/ classmethods, right?), but `lru_cache` worked just fine. That said, it is about 50% slower.
  - current implementation at 100k runs takes ~.28s on average.
  - `lru_cache` at 100k runs takes ~.42s on average.

50% is a significant percentage, but we're talking about lookup resolution taking .0000042 seconds vs .0000028 seconds. `¯\_(ツ)_/¯`

Either way, I've squashed the existing commits into one, and added the lru changes separately. Thoughts on the performance vs code clarity?

timgraham commented:
I don't have a sense as to whether or not those performance differences would have any real-world impact.

If we keep the changes, I think you can drop `maxsize=None` since the methods don't have any arguments, the size shouldn't be larger than 1, right?

rpkilby commented:
> I don't have a sense as to whether or not those performance differences would have any real-world impact.

Yeah. With my laptop, it take 700,000 lookups to use one additional second of CPU time. I feel like this is probably fine. Went ahead and squashed the commits.

> If we keep the changes, I think you can drop maxsize=None since the methods don't have any arguments, the size shouldn't be larger than 1, right?

No - it should be caching per field class. eg, the datetime lookups vs the standard lookups.
